### PR TITLE
Registry: Relax rules to allow valid stakes to register

### LINKED ISSUE #2524: Registry: Relax rules to allow valid stakes to register
Currently, only enrolled validators may register. However, this means that nodes are not immediately reachable when they enroll, and there is a delay in between the moment a block is externalized and nodes can get an address to contact those new peers.
If the block interval is too low, or one's retry delay too high, it means a node might miss one or more consensus rounds.

### PR DESCRIPTION
**Pull request by omerfirmak**
Builds on top of #2487

Fixes #2524

### CODE REVIEW

**Comment by omerfirmak** [comment]
Alternative to commit#4 is to reject any validators that can not prove a stake, I am not completely sure which way is the right way to go. Feel free to comment.

**Comment by omerfirmak** [comment]
Added the fix for #2576

File: source/agora/test/Base.d [hunk 1]
```diff
@@ -2234,7 +2234,7 @@ public APIManager makeTestNetwork (APIManager : TestAPIManager = TestAPIManager)
         .map!(en => validatorAddress(en.index, en.value)).array;
 
     auto full_node_addresses = test_conf.full_nodes.iota.map!(
-        idx => fullNodeAddress(idx)).array;
+        idx => fullNodeAddress(idx)).array ~ ["name.registry"];
```
Comment (Geod24) [inline source/agora/test/Base.d:2237]: Doesn't seem to belong to this commit ?

File: source/agora/node/Registry.d [hunk 2]
```diff
@@ -210,8 +210,14 @@ public class NameRegistry: NameRegistryAPI
             this.validator_info = this.ledger.getValidators(last_height);
             this.validator_info_height = last_height;
         }
+        UTXO utxo;
         ensure(this.validator_info.map!(info => info.address)
-            .canFind(registry_payload.data.public_key), "Not an enrolled validator");
+            .canFind(registry_payload.data.public_key) ||
+            this.ledger.enrollment_manager.enroll_pool
+                .getEnrollments(last_height).canFind!(
```
Comment (Geod24) [inline source/agora/node/Registry.d]: Why do we check the enrollment pool ?

File: source/agora/consensus/Ledger.d [hunk 3]
```diff
@@ -85,7 +85,7 @@ public class Ledger
     private Block last_block;
 
     /// UTXO set
-    private UTXOCache utxo_set;
+    public UTXOCache utxo_set;
```
Comment (Geod24) [inline source/agora/consensus/Ledger.d]: Just move `ValidatingLedger.peekUTXO` in `Ledger` and you can avoid exposing internals

File: source/agora/node/Registry.d [hunk 4]
```diff
@@ -198,8 +197,7 @@ public class NameRegistry: NameRegistryAPI
 
     public override void postValidator (RegistryPayload registry_payload)
     {
-        import std.algorithm;
-        import agora.consensus.data.UTXO;
+        import std.range;
```
Comment (Geod24) [inline source/agora/node/Registry.d]: Move to top level

**Review by Geod24 — CHANGES_REQUESTED** [review]

File: source/agora/node/Registry.d (hunk 2 above)
Comment (omerfirmak) [reply source/agora/node/Registry.d]: IIRC, that was how we discussed to go about this initially.

This way we don't need to do any checks on the frozen UTXO. If the enrollment is in the pool already, we know for sure that it satisfies all the requirements.

Comment (Geod24) [reply source/agora/node/Registry.d]: But we still risk a race condition if the node's `Enrollment` doesn't make it to the registry.
Of course this can be trivially worked around by the client by sending the `Enrollment` first, but it looks like an un-needed edge case.

Comment (Geod24) [reply source/agora/node/Registry.d]: I guess as a first step it would be fine, though, if the test suite goes green with this approach.

Comment (omerfirmak) [reply source/agora/node/Registry.d]: All green apart from the weird hang in between tests.

**Review by Geod24 — APPROVED** [review]
